modules/core/src/main/scala/iota/internal/toolbelts.scala

```diff
@@ -0,0 +1,166 @@
+        if (showCache)
+          c.echo(c.enclosingPosition, s"ShowCache: $b cached result $b")
+        b
+      case _ =>
+        val b = f(a)
+        cache.underlying += ((a, b))
+        if (showCache)
+          c.echo(c.enclosingPosition, s"ShowCache: $a computed result $b")
+        b
+    }
+  }
+
+  def memoizedTListTypes(tpe: Type): Either[Id[String], List[Type]] =
+    memoize(IotaMacroToolbelt.typeListCache)(tpe, tlistTypes)
+
+  def memoizedKListTypes(tpe: Type): Either[Id[String], List[Type]] =
+    memoize(IotaMacroToolbelt.typeListCache)(tpe, klistTypes)
+
+}
+
+object IotaReflectiveToolbelt {
+  def apply[U <: Universe](u: U): IotaReflectiveToolbelt[u.type] =
+    new IotaReflectiveToolbelt[u.type](u)
+
+  def apply(): IotaReflectiveToolbelt[runtimeUniverse.type] =
+    apply(runtimeUniverse)
+}
+
+class IotaReflectiveToolbelt[U <: Universe](override val u: U)
+    extends IotaCommonToolbelt { override type Uu = U }
+
+private[internal] sealed abstract class IotaCommonToolbelt {
+  type Uu <: Universe
+  val u: Uu
+
+  import u._
+
+  final lazy val TNilTpe  = typeOf[iota.TNil]
+  final lazy val TConsTpe = typeOf[iota.TCons[Nothing, Nothing]].etaExpand.resultType
+
+  final lazy val KNilTpe  = typeOf[iota.KNil]
+  final lazy val KConsTpe = typeOf[iota.KCons[Nothing, Nothing]].etaExpand.resultType
+
+  final lazy val CopTpe   = typeOf[iota.Cop[Nothing]].etaExpand.resultType
+  final lazy val CopKTpe  = typeOf[iota.CopK[Nothing, Nothing]].etaExpand.resultType
+
+  @tailrec
+  final def typeListFoldLeft[A](
+    nilTpe: Type, consTpe: Type
+  )(tpe: Type)(a0: A)(f: (A, Type) => A): Either[String, A] = tpe.dealias match {
+    case TypeRef(_, sym, Nil) if sym.asType.toType <:< nilTpe => Right(a0)
+    case TypeRef(_, sym, head :: tail :: Nil) if sym.asType.toType <:< consTpe =>
+      typeListFoldLeft(nilTpe, consTpe)(tail)(f(a0, head))(f)
+    case _ =>
+      Left(s"Unexpected type ${showRaw(tpe)} when inspecting type list")
+  }
+
+  final def typeListTypes(
+    nilTpe: Type, consTpe: Type
+  )(tpe: Type): Either[Id[String], List[Type]] =
+    typeListFoldLeft(nilTpe, consTpe)(tpe)(List.empty[Type])((acc, t) => t :: acc).map(_.reverse)
+
+  final def tlistTypes(tpe: Type): Either[Id[String], List[Type]] =
+    typeListTypes(TNilTpe, TConsTpe)(tpe)
+
+  final def klistTypes(tpe: Type): Either[Id[String], List[Type]] =
+    typeListTypes(KNilTpe, KConsTpe)(tpe)
+
+  final def klistTypeConstructors(tpe: Type): Either[Id[String], List[Type]] =
+    klistTypes(tpe).map(_.map(_.etaExpand.resultType))
+
+  case class CopTypes(L: Type)
+  case class CopKTypes(L: Type, A: Type)
+
+  private[this] def resultType(sym: Symbol): Type =
+    sym.asType.toType.etaExpand.resultType
+
+  final def destructCop(tpe: Type): Either[Id[String], CopTypes] =
+    tpe.dealias match {
+      case TypeRef(_, sym, l :: Nil) if resultType(sym) <:< CopTpe => Right(CopTypes(l))
+      case TypeRef(_, sym, Nil) => destructCop(sym.asType.toType)
```

Comment:
Do we not have `sym.typeSignature` assuming this is a `TypeSymbol`? `sym.asType.toType` can throw `scala.ScalaReflectionException` as far as I know.